streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollector.java

```diff
@@ -64,9 +68,20 @@ public RecordCollector(Producer<byte[], byte[]> producer) {
     }
 
     public <K, V> void send(ProducerRecord<K, V> record, Serializer<K> keySerializer, Serializer<V> valueSerializer) {
+        send(record, keySerializer, valueSerializer, null);
+    }
+
+    public <K, V> void send(ProducerRecord<K, V> record, Serializer<K> keySerializer, Serializer<V> valueSerializer,
+                            StreamPartitioner<K, V> partitioner) {
         byte[] keyBytes = keySerializer.serialize(record.topic(), record.key());
         byte[] valBytes = valueSerializer.serialize(record.topic(), record.value());
-        this.producer.send(new ProducerRecord<>(record.topic(), keyBytes, valBytes), callback);
+        Integer partition = null;
+        if (partitioner != null) {
+            List<PartitionInfo> partitions = this.producer.partitionsFor(record.topic());
+            if (partitions != null)
+                partition = partitioner.partition(record.key(), record.value(), partitions.size());
```

Comment:
```KafkaProducer.partitionsFor()``` could block up to max-block-time when the metadata of the topic does not exist, so if it returns null it means the producer cannot fetch the metadata within the block time period, which should not be common. Also the same function will be called in ```send()``` and if it returns null we treat it as an unexpected error.